library.sh

```diff
@@ -429,45 +434,46 @@ function mktemp_with_extension() {
 function create_junit_xml() {
   local xml
   xml="$(mktemp_with_extension "${ARTIFACTS}"/junit_XXXXXXXX xml)"
-  echo "JUnit file ${xml} is created for reporting the test result"
+  echo "XML report for $1::$2 written to ${xml}"
   run_kntest junit --suite="$1" --name="$2" --err-msg="$3" --dest="${xml}" || return 1
 }
 
 # Runs a go test and generate a junit summary.
 # Parameters: $1... - parameters to go test
 function report_go_test() {
   local go_test_args=( "$@" )
-  # Install gotestsum if necessary.
-  run_go_tool gotest.tools/gotestsum gotestsum --help > /dev/null 2>&1
-  # Capture the test output to the report file.
-  local report
-  report="$(mktemp)"
-  local xml
+  local logfile xml ansilog htmllog
   xml="$(mktemp_with_extension "${ARTIFACTS}"/junit_XXXXXXXX xml)"
+  # Keep the suffix, so files are related.
+  logfile="${xml/junit_/go_test_}"
+  logfile="${logfile/.xml/.jsonl}"
   echo "Running go test with args: ${go_test_args[*]}"
-  capture_output "${report}" gotestsum --format "${GO_TEST_VERBOSITY:-testname}" \
-    --junitfile "${xml}" --junitfile-testsuite-name relative --junitfile-testcase-classname relative \
+  go_run gotest.tools/gotestsum@v1.8.0 \
+    --format "${GO_TEST_VERBOSITY:-testname}" \
+    --junitfile "${xml}" \
+    --junitfile-testsuite-name relative \
+    --junitfile-testcase-classname relative \
+    --jsonfile "${logfile}" \
```

Comment:
Do the `junitfile` and `jsonfile` have different contents? If not, I'd prefer to collect one (whichever has more information), and then convert from one to the other later if necessary. Cloud storage may be cheap, but it's not completely free, and keeping two copies of stuff that we may never look at seems excessive.

Reply:
Yes, they differ.

The junit file contains the XML with test names and execution time. If test is successful, this XML doesn't contain the test output if the test is successful. See, [example file](https://storage.googleapis.com/knative-prow/pr-logs/pull/knative_hack/172/unit-tests_hack_main/1519415863650816000/artifacts/junit_xUZhYeE4.xml). This file is read by Prow as well.

The json file contains a JSON object for every line of output the tests have produced, including the Golang output and the output of the successful tests as well. This can be handy for integration tests and upgrade tests. For both of them, the real reason for failure in some step might be hidden in the output of previous successful `t.Run` subtest - for example, an WARN log message. The idea is to archive both of them, in case the debugging is needed.